File: iOS7bookExamples/bk2ch06p324containerViewControllerStoryboard/ch19p642containerViewControllerStoryboard/ViewController2.m


#import "ViewController2.h"

@interface ViewController2 () <UIViewControllerTransitioningDelegate, UIViewControllerAnimatedTransitioning>

@end

@implementation ViewController2

-(void)awakeFromNib {
    [super awakeFromNib];
    self.transitioningDelegate = self;
    self.modalPresentationStyle = UIModalPresentationCurrentContext;
}
- (IBAction)doButton:(id)sender {
    [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view.
}

-(id<UIViewControllerAnimatedTransitioning>)animationControllerForPresentedController:(UIViewController *)presented presentingController:(UIViewController *)presenting sourceController:(UIViewController *)source {
    return self;
}

-(NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext {
    return 0.8;
}

-(void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext {
    // UIViewController* vc1 = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
    UIViewController* vc2 = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
    UIView* con = [transitionContext containerView];
    // CGRect r1start = [transitionContext initialFrameForViewController:vc1];
    CGRect r2end = [transitionContext finalFrameForViewController:vc2];
    // UIView* v1 = vc1.view;
    UIView* v2 = vc2.view;

    CGRect r2start = r2end;
    r2start.origin.y -= r2start.size.height;
    v2.frame = r2start;
    [con addSubview:v2];
    
    [UIView animateWithDuration:0.8 delay:0 usingSpringWithDamping:0.5 initialSpringVelocity:20 options:0 animations:^{
        v2.frame = r2end;
    } completion:^(BOOL finished) {
        [transitionContext completeTransition:YES];
    }];
    
}


@end

File: iOS9bookExamples/bk2ch06p324containerViewControllerStoryboard/ch19p642containerViewControllerStoryboard/ViewController2.swift
import UIKit

class ViewController2 : UIViewController {
    @IBAction func doButton(sender:AnyObject?) {
        self.presentingViewController!.dismissViewControllerAnimated(true, completion: nil)
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        print("vc2 view did load")
    }
}

// ===== the rest of the example actually demonstrates something else entirely,
// namely a current-context presentation (with custom animation)

extension ViewController2 {
    override func awakeFromNib() {
        super.awakeFromNib()
        self.transitioningDelegate = self
        self.modalPresentationStyle = .CurrentContext
    }
}

extension ViewController2 : UIViewControllerTransitioningDelegate {
    func animationControllerForPresentedController(presented: UIViewController, presentingController presenting: UIViewController, sourceController source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        return self
    }
}

extension ViewController2 : UIViewControllerAnimatedTransitioning {
    func transitionDuration(transitionContext: UIViewControllerContextTransitioning?) -> NSTimeInterval {
        return 0.8
    }
    
    func animateTransition(transitionContext: UIViewControllerContextTransitioning) {
        let vc2 = transitionContext.viewControllerForKey(UITransitionContextToViewControllerKey)!
        
        let con = transitionContext.containerView()!
        
        let r2end = transitionContext.finalFrameForViewController(vc2)
        
        let v2 = transitionContext.viewForKey(UITransitionContextToViewKey)!
                
        var r2start = r2end
        r2start.origin.y -= r2start.size.height
        v2.frame = r2start
        con.addSubview(v2)
        UIView.animateWithDuration(0.8, delay: 0, usingSpringWithDamping: 0.5, initialSpringVelocity: 10, options: [], animations: {
            v2.frame = r2end
            }, completion: {
                _ in
                transitionContext.completeTransition(true)
            })

    }
}